controllers/vspherepolicy/policyevaluation/policyevaluation_controller.go

```diff
@@ -0,0 +1,583 @@
+	if err != nil {
+		return false, fmt.Errorf("failed to match image labels: %w", err)
+	}
+
+	return labelSelector.Matches(labels.Set(obj.Spec.Image.Labels)), nil
+}
+
+func evaluateImageNameMatch(
+	obj *vspherepolv1.PolicyEvaluation,
+	nameSpec *vspherepolv1.StringMatcherSpec) (bool, error) {
+
+	if nameSpec == nil {
+		return true, nil
+	}
+
+	if obj.Spec.Image == nil {
+		return false, nil
+	}
+
+	return matchesString(obj.Spec.Image.Name, nameSpec)
+}
+
+func matchesString(
+	actual string,
+	matcher *vspherepolv1.StringMatcherSpec) (bool, error) {
+
+	if matcher == nil {
+		return true, nil
+	}
+
+	switch matcher.Op {
+	case vspherepolv1.ValueSelectorOpEqual, "":
+		return actual == matcher.Value, nil
+	case vspherepolv1.ValueSelectorOpNotEqual:
+		return actual != matcher.Value, nil
+	case vspherepolv1.ValueSelectorOpHasPrefix:
+		return strings.HasPrefix(actual, matcher.Value), nil
+	case vspherepolv1.ValueSelectorOpNotHasPrefix:
+		return !strings.HasPrefix(actual, matcher.Value), nil
+	case vspherepolv1.ValueSelectorOpHasSuffix:
+		return strings.HasSuffix(actual, matcher.Value), nil
+	case vspherepolv1.ValueSelectorOpNotHasSuffix:
+		return !strings.HasSuffix(actual, matcher.Value), nil
+	case vspherepolv1.ValueSelectorOpContains:
+		return strings.Contains(actual, matcher.Value), nil
+	case vspherepolv1.ValueSelectorOpNotContains:
+		return !strings.Contains(actual, matcher.Value), nil
+	case vspherepolv1.ValueSelectorOpMatch:
+		return regexp.MatchString(matcher.Value, actual)
+	case vspherepolv1.ValueSelectorOpNotMatch:
+		ok, err := regexp.MatchString(matcher.Value, actual)
+		return !ok, err
+	default:
+		return false, nil
+	}
+}
+
+func matchesGuestFamily(
+	actual vspherepolv1.GuestFamilyType,
+	matcher *vspherepolv1.GuestFamilyMatcherSpec) (bool, error) {
+
+	var sm *vspherepolv1.StringMatcherSpec
+	if matcher != nil {
+		sm = &vspherepolv1.StringMatcherSpec{
+			Op:    matcher.Op,
+			Value: string(matcher.Value),
+		}
+	}
+
+	return matchesString(string(actual), sm)
+}
+
+const computePolicyKind = "ComputePolicy"
+
+func (r *Reconciler) reconcileExplicitPolicies(
+	ctx context.Context,
+	obj *vspherepolv1.PolicyEvaluation) error {
+
+	for _, ref := range obj.Spec.Policies {
+		switch ref.Kind {
+		case computePolicyKind:
+			if err := r.addComputePolicyRef(ctx, obj, ref); err != nil {
+				return fmt.Errorf(
+					"failed to add explicit compute policy %s: %w",
+					ref.Name, err)
+			}
+		default:
+			// Log and skip unknown policy kinds
+			r.Logger.Info("skipping unknown policy kind",
+				"kind", ref.Kind,
+				"name", ref.Name)
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) addComputePolicyRef(
+	ctx context.Context,
+	obj *vspherepolv1.PolicyEvaluation,
+	ref vspherepolv1.LocalObjectRef) error {
+
+	var (
+		pol vspherepolv1.ComputePolicy
+		key = client.ObjectKey{
+			Namespace: obj.Namespace,
+			Name:      ref.Name,
+		}
+	)
+
+	if err := r.Client.Get(ctx, key, &pol); err != nil {
```

Comment:
All ComputePolicys are fetched  fetched in `reconcileMatchingComputePolicies`. Maybe you can extract the List on the top level and store in a map then pass to addComputePolicyRef so that you don't need to fetch it again?

Reply:
Yeah I forgot about the client-go cache again 😂